creating migration for invoice,invoice_products and invoice_payment a…

HamidNiakan opened the pull request:
…nd then creating class as InvoiceManager for CRUD

HamidNiakan committed:
creating migration for invoice,invoice_products and invoice_payment and then creating class as InvoiceManager for CRUD

HamidNiakan committed:
config unitTest and creating class as TestCase,create UserFactory, and also creating config for package as invoice.php

HamidNiakan committed:
creating UnitTest for testing InvoiceManager class and fix bug in migrattion and model and also creating Exception As AmountInvoiceException

HamidNiakan committed:
creating factory for invoice,product and payment model,refactoring InvoiceManagerTest, remvoing extra code in TestCase,creating exception as InvoiceUserMismatchException, and then creating UserFactory for generate User fake

HamidNiakan committed:
run fixer

HamidNiakan committed:
testing AddProductToInvoce and some change for coverge 100

HamidNiakan committed:
chnaging phpunit.xml and gitignore

yeganemehr commented on database/migrations/2023_01_18_000002_create_invoices_products_table.php:
```diff
@@ -0,0 +1,42 @@
+             $table->timestamps();
+             $table->decimal('price', 10 + $floatScale, $floatScale);
+             $table->decimal('discount', 10 + $floatScale, $floatScale)->nullable();
+             $table->decimal('total_amount', 10 + $floatScale, $floatScale)->nullable();
```
calculate this in your model and don't save in it in the table.

HamidNiakan replied:
done

yeganemehr commented on database/migrations/2023_01_18_000002_create_invoices_products_table.php:
```diff
@@ -0,0 +1,42 @@
+             $table->json('meta')->nullable();
+             $table->json('distribution_plan')->nullable();
+             $table->json('distribution')->nullable();
+             $table->longText('description')->nullable();
```
use text. descriptions in invoice items are not that long

HamidNiakan replied:
done

yeganemehr commented on database/migrations/2023_01_18_000002_create_invoices_products_table.php:
```diff
@@ -0,0 +1,42 @@
+             $table->json('distribution')->nullable();
+             $table->longText('description')->nullable();
+
+             $table->foreign('Invoice_id')
```
merge these `foreign()`s with above `foreignId()`s

HamidNiakan replied:
done

yeganemehr commented on config/invoice.php:
```diff
@@ -0,0 +1,6 @@
+<?php
+
+return [
+    // Define your user model class for connect tickets to users.
```
tickets?

HamidNiakan replied:
done

yeganemehr commented on database/factories/InvoiceFactory.php:
```diff
@@ -0,0 +1,125 @@
+                                           ->asEUR());
+    }
+
+    public function paid()
```
when it's paid, it's must has paid_time

HamidNiakan replied:
done

yeganemehr commented on database/migrations/2023_01_18_000003_create_invoices_payments_table.php:
```diff
@@ -0,0 +1,37 @@
+             $table->timestamps();
+             $table->string('method');
+             $table->decimal('amount', 10 + $floatScale, $floatScale);
+             $table->string('status');
```
use enums

HamidNiakan replied:
done

yeganemehr commented on database/migrations/2023_01_18_000003_create_invoices_payments_table.php:
```diff
@@ -0,0 +1,37 @@
+
+     public function up(): void
+     {
+         Schema::create('invoice_payments', function (Blueprint $table) {
```
```suggestion
         Schema::create('invoices_payments', function (Blueprint $table) {
```

HamidNiakan replied:
done

yeganemehr commented on composer.json:
```diff
@@ -13,7 +14,9 @@
-		"dnj/laravel-account": "dev-master"
+		"dnj/laravel-account": "dev-master",
+		"dnj/laravel-currency": "^1.0.1",
+		"laravel/legacy-factories": "^1.3"
```
why did you need legacy-factories?

HamidNiakan replied:
This was for the previous changes I made to the User Factory
done

yeganemehr commented on database/factories/PaymentFactory.php:
```diff
@@ -0,0 +1,70 @@
+        return [
+            'invoice_id' => Invoice::factory(),
+            'transaction_id' => null,
+            'method' => fake()->sentence(3),
```
it's should be class name which implemented `Contracts\IPaymentMethod` interface

HamidNiakan replied:
I can't implement it

yeganemehr commented on database/factories/ProductFactory.php:
```diff
@@ -0,0 +1,123 @@
+        ]);
+    }
+
+    public function withTile(string $title)
```
```suggestion
    public function withTitle(string $title)
```

HamidNiakan replied:
done

yeganemehr commented on database/factories/ProductFactory.php:
```diff
@@ -0,0 +1,123 @@
+        ]);
+    }
+
+    public function withInvoice(int $invoiceId)
```
Let's accept Invoice|InvoiceFactory instance instead of just id.

HamidNiakan replied:
done

yeganemehr commented on src/InvoiceManager.php:
```diff
@@ -0,0 +1,326 @@
+            throw new InvalidInvoiceStatusException();
+        }
+        if ($payment->getAmount() > $payment->invoice->getAmount()) {
+            throw new AmountInvoiceMismatchException();
```
You check amount of payment after you inserted in the database?

HamidNiakan replied:
done

yeganemehr commented on src/InvoiceManager.php:
```diff
@@ -0,0 +1,326 @@
+        }
+        if ($payment->getAmount() == $payment->invoice->getAmount()) {
+            $payment->update([
+                                 'status' => PaymentStatus::APPROVED,
```
you just approve payment just when client pay invoice in one part?

HamidNiakan replied:
done

yeganemehr commented on src/InvoiceManager.php:
```diff
@@ -0,0 +1,326 @@
+        if (InvoiceStatus::PAID == $payment->invoice->getStatus()) {
+            throw new InvalidInvoiceStatusException();
+        }
+        $payment->update([
```
What if payment already approved? no need to check status of payment?

HamidNiakan replied:
done

yeganemehr commented on src/InvoiceManager.php:
```diff
@@ -0,0 +1,326 @@
+        // TODO: Implement getById() method.
+        return Invoice::query()
+                      ->findOrFail($invoiceId)
+                      ->load('products');
```
if you want to [eager loading](https://laravel.com/docs/9.x/eloquent-relationships#eager-loading), put it in the model.

HamidNiakan replied:
done

yeganemehr commented on src/InvoiceManager.php:
```diff
@@ -0,0 +1,326 @@
+        }
+        $payment = $invoice->payments()
+                           ->create([
+                                        'method' => $type,
```
Did you check type is valid class name that is implemented `IPaymentMethod` interface?

HamidNiakan replied:
I can't implement it

yeganemehr commented on src/InvoiceManager.php:
```diff
@@ -0,0 +1,326 @@
+        $invoice->user_id = $userId;
+        $invoice->currency_id = $currencyId;
+        $invoice->meta = $meta;
+        if (isset($localizedDetails['title'])) {
```
title is not optional

HamidNiakan replied:
done

yeganemehr commented on src/InvoiceManager.php:
```diff
@@ -0,0 +1,326 @@
+        if (isset($localizedDetails['title'])) {
+            $invoice->title = $localizedDetails['title'];
+        }
+        $invoice->status = InvoiceStatus::UNPAID->value;
```
```suggestion
        $invoice->status = InvoiceStatus::UNPAID;
```

HamidNiakan replied:
done

yeganemehr commented on src/InvoiceManager.php:
```diff
@@ -0,0 +1,326 @@
+        $invoice->save();
+        $invoice->products()
+                ->createMany($this->buliderInvoiceProducts($products));
+        $invoice->update([
```
update after creation? :exploding_head:

HamidNiakan replied:
done

yeganemehr commented on src/InvoiceManager.php:
```diff
@@ -0,0 +1,326 @@
+            $invoice->payments()
+                    ->delete();
+        }
+        $invoice->delete();
```
just make relations cascade on delete and delete invoice. no need to delete payments and products separately.

HamidNiakan replied:
done

yeganemehr commented on src/InvoiceManager.php:
```diff
@@ -0,0 +1,326 @@
+
+        return DB::transaction(function () use ($invoice, $product) {
+            $record = new Product();
+            if (isset($product['price'])) {
```
can we add a product to the invoice without price or title or count or ...?

HamidNiakan replied:
done

yeganemehr commented on src/ModelHelpers.php:
```diff
@@ -0,0 +1,35 @@
+
+    protected function getCurrencyTable(): string
+    {
+        return (new Currency())->getTable();
```
You can use Currency model statically

HamidNiakan replied:
done

yeganemehr commented:
There is MUCH to change
I just got tired of this
waste of time

HamidNiakan commented on database/migrations/2023_01_18_000001_create_invoices_table.php:
```diff
@@ -0,0 +1,46 @@
+             $table->decimal('paid_amount', 10 + $floatScale, $floatScale)->nullable();
+             $table->decimal('unpaid_amount', 10 + $floatScale, $floatScale)->nullable();
+             $table->json('meta')->nullable();
+             $table->dateTime('paid_time')->nullable();
```
done

HamidNiakan commented on database/migrations/2023_01_18_000001_create_invoices_table.php:
```diff
@@ -0,0 +1,46 @@
+             $table->timestamps();
+             $table->decimal('amount', 10 + $floatScale, $floatScale)->nullable();
+             $table->decimal('paid_amount', 10 + $floatScale, $floatScale)->nullable();
+             $table->decimal('unpaid_amount', 10 + $floatScale, $floatScale)->nullable();
```
done

HamidNiakan commented on src/Models/Invoice.php:
```diff
@@ -150,17 +155,22 @@ public function recalculateTotalAmount(): void
             $amountInSameCurrency = $exchange->convert($product->getTotalAmount(), $this->getCurrencyId(), $product->getCurrencyId(), true);
             $result = $result->add($amountInSameCurrency);
         }
+        dd($result);
```
done

HamidNiakan commented on tests/Unit/InvoiceManagerTest.php:
```diff
@@ -0,0 +1,429 @@
+            'currencyId' => $EUR->getID(),
+            'products' => [
+                [
+                    'id' => 1,
```
done

HamidNiakan commented on tests/Unit/InvoiceManagerTest.php:
```diff
@@ -0,0 +1,429 @@
+        $now = time();
+        $user = User::factory()
+                    ->create();
+        $USD = $this->createUSD();
```
done

HamidNiakan commented on database/migrations/2023_01_18_000002_create_invoices_products_table.php:
```diff
@@ -0,0 +1,42 @@
+
+     public function up(): void
+     {
+         Schema::create('invoice_products', function (Blueprint $table) {
```
done

HamidNiakan commented on database/migrations/2023_01_18_000001_create_invoices_table.php:
```diff
@@ -0,0 +1,46 @@
+             $table->decimal('unpaid_amount', 10 + $floatScale, $floatScale)->nullable();
+             $table->json('meta')->nullable();
+             $table->dateTime('paid_time')->nullable();
+             $table->string('status')->nullable();
```
done

HamidNiakan commented on database/migrations/2023_01_18_000001_create_invoices_table.php:
```diff
@@ -0,0 +1,46 @@
+             $table->dateTime('paid_time')->nullable();
+             $table->string('status')->nullable();
+
+             $table->foreign('currency_id')
```
done

HamidNiakan committed:
changing tables name
changing enums for Invoice and Payment and move to directory as enum
removing extra field in table invoices and
fix up function addProductToInvoice in InvoiceManager
fix up function delete in  InvoiceManager
removing extra package in composer.json
fix up  function paid in InvoiceFactory

HamidNiakan committed:
upgrading laravel-accounting package
adding field in invoice payments table as currency id
rebuilding functions Add Payment To Invoice,Approve Payment,Rejected Payment
creating exception as FinishedInvoicePaymentsException
remoing extra code and refactoring codes

HamidNiakan committed:
remvoing getCurrencyTable and set table name statically

HamidNiakan committed:
fix code style and invoice config

yeganemehr committed:
Refactor

yeganemehr committed:
add github workflow for running tests and check code coverage
